### ISSUE TITLE
Template tag for EdgeQL

### ISSUE THREAD

**Issue by colinhacks** [edgedb/edgedb-js#350]
Proposal: provide a template tag for writing sanitized EdgeQL queries.

### User input

```typescript
const query = edgeql`insert User { title := ${req.body.title}`
// => { query: 'insert User { title := <str>$var_0 }', parameters: {var_0: "Some title"}}
```

We'd overload the query methods to support an object-based query format: `{query: string; parameters: {[k:string]: any}}`

```ts
await client.query(query)
```

The cast type will be inferred from the input data according to the following rules:

- `Array.isArray(data) === true` -> check that contents are homogenous primitives -> `array<element-type>` 
- `object` -> `json`
- `string` -> `str`
- `number` -> `float64`
- `bigint` -> `bigint`
- `boolean` -> `bool`

### Identifiers

There should be a separate API for specifying identifiers: `edgeql.id()`. This would verify that the input is a valid identifier with a regex.

```ts
const field1 = 'name'
const field2 = 'email'
edgeql`insert User { ${edgeql.id(field1)} ${edgeql.id(field2)} }`
```

For convenience this function can be variadic:

```ts
edgeql`insert User { ${edgeql.id(field1, field2)} }`
```

### Fluent API

```
### Prior art:

https://github.com/felixfbecker/node-sql-template-strings
https://github.com/blakeembrey/sql-template-tag
https://www.npmjs.com/package/pg-format

**Comment by 1st1** [comment]
> For convenience this function can be variadic. The inputs will be validated and joined with ", ".

Probably it accepting an array of strings is better since I imagine sometimes such lists would be formed dynamically, and `ident(list)` is better than `ident.apply(null, list)`.

**Comment by colinhacks** [comment]
> Probably it accepting an array of strings is better

For arrays you'd use the spread operator: `ident(...list)`

**Comment by elprans** [comment]
> This would verify that the input is a valid identifier with a regex.

Pretty much any Unicode string is a valid identifier as long as you quote it.  The only hard rule is that an ident cannot start with `$` or `@`.

**Comment by jaclarke** [comment]
> The cast type will be inferred from the input data according to the following rules:

I think it's probably safe to add the rest of the types to this list, ie. datetime, local_date/time, duration and bytes, since we can identify them with `instanceof`. The only exception would then be `uuid`, and the user would just have manually add a cast.

**Comment by colinhacks** [comment]
> I think it's probably safe to add the rest of the types to this list

Indeed good call!

**Comment by infogulch** [comment]
I see draft PR #381 implements this functionality. See also [discussion on edgedb/edgedb](https://github.com/edgedb/edgedb/discussions/4051).

This may be premature, but is there a way to expose the result type of a tagged template query to typescript for autocompletion / typechecking purposes? E.g. using `.name` on an entry of a select query that does not include a `name` field should be a type error.

**Comment by colinhacks** [comment]
No, this template tag will not be schema-aware nor can it infer types. That's virtually impossible to implement with just template literals - you should use the query builder if you want inferred types and typechecking.

**Comment by haikyuu** [comment]
> That's virtually impossible to implement with just template literals
Indeed, but it would be possible with a VSCode extension. It would introspect each query at dev-time and tell typescript what type that is

**Comment by scotttrinh** [comment]
Is this still desirable given the `queries` generator gives us something pretty close to this experience without needing to jump through a bunch of tooling hoops to get good type information? If the desired outcome was just sanitized inputs rather than co-location, feels like the `queries` generator fits the bill.
